crates/kitsune2/tests/integration.rs

```diff
@@ -0,0 +1,350 @@
+                actual_ops_2.len(),
+                ops_1.len()
+            );
+        }
+    });
+}
+
+/// Test that space shutdown is reasonably clean:
+/// - Start two Kitsune2 instances
+/// - Record the initial number of Tokio tasks
+/// - Start two spaces, one on each instance
+/// - Join an agent to each space
+/// - Create some ops in each space
+/// - Wait for gossip to exchange all ops
+/// - Have local agents leave the spaces
+/// - Wait for all peers to declare a tombstone
+/// - Shut down the spaces
+/// - Wait for the spaces' tasks to be cleaned up
+///
+/// This isn't a perfect check for shutdown, but it's a reasonable expectation that if all the
+/// Tokio tasks for a space are gone, then it's not actively doing work in the background.
+#[tokio::test]
+async fn shutdown_space() {
+    enable_tracing();
+
+    let signal_server = SbdServer::new(Arc::new(sbd_server::Config {
+        bind: vec!["127.0.0.1:0".to_string()],
+        ..Default::default()
+    }))
+    .await
+    .unwrap();
+    let signal_server_url = format!("ws://{}", signal_server.bind_addrs()[0]);
+
+    let bootstrap_server = TestBootstrapSrv::new(false).await;
+    let bootstrap_server_url = bootstrap_server.addr().to_string();
+
+    // Create 2 Kitsune instances..
+    let kitsune_1 =
+        make_kitsune_node(&signal_server_url, &bootstrap_server_url).await;
+    let kitsune_2 =
+        make_kitsune_node(&signal_server_url, &bootstrap_server_url).await;
+
+    let metrics = tokio::runtime::Handle::current().metrics();
+    let initial_tasks = metrics.num_alive_tasks();
+
+    // and 1 space with 1 joined agent each.
+    let space_1 = start_space(&kitsune_1).await;
+    let space_2 = start_space(&kitsune_2).await;
+
+    // Create some data for each agent
+    let (ops_1, op_ids_1) = create_op_list(10);
+    space_1
+        .op_store()
+        .process_incoming_ops(ops_1.clone())
+        .await
+        .unwrap();
+    let (ops_2, op_ids_2) = create_op_list(10);
+    space_2
+        .op_store()
+        .process_incoming_ops(ops_2.clone())
+        .await
+        .unwrap();
+
+    // Wait for gossip to exchange all ops.
+    iter_check!(15000, 500, {
+        let actual_ops_1 = space_1
+            .op_store()
+            .retrieve_ops(op_ids_2.clone())
+            .await
+            .unwrap();
+        let actual_ops_2 = space_2
+            .op_store()
+            .retrieve_ops(op_ids_1.clone())
+            .await
+            .unwrap();
+        if actual_ops_1.len() == ops_2.len()
+            && actual_ops_2.len() == ops_1.len()
+        {
+            break;
+        } else {
+            println!(
+                "space 1 actual ops received {}/expected {}",
+                actual_ops_1.len(),
+                ops_2.len()
+            );
+            println!(
+                "space 2 actual ops received {}/expected {}",
+                actual_ops_2.len(),
+                ops_1.len()
+            );
+        }
+    });
+
```

Comment:
My comment didn't come through here yesterday. I suggest to remove the space here and assert it fails, because there are still local agents connected.

Reply:
Good call, done https://github.com/holochain/kitsune2/pull/208/commits/7652800ce339ca77a0b7a127bafbeedf021ad092